File: src/dungeonbreak_narrative/escape_the_dungeon/narrative/prerequisites.py
"""Reusable prerequisite checks for actions, dialogue, and skills."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Iterable, Sequence

from ..entities.models import EntityState, ItemInstance
from ..world.map import RoomNode


@dataclass(frozen=True)
class Prerequisite:
    kind: str
    key: str = ""
    value: str | float | int | None = None
    description: str = ""


@dataclass(frozen=True)
class AvailabilityResult:
    available: bool
    blocked_reasons: tuple[str, ...] = ()


@dataclass
class PrerequisiteContext:
    actor: EntityState
    room: RoomNode
    nearby_entities: Sequence[EntityState] = ()
    target: EntityState | None = None
    exits: Sequence[str] = ()
    extra_flags: dict[str, bool] = field(default_factory=dict)


def _inventory_has_tag(items: Iterable[ItemInstance], tag: str) -> bool:
    lowered = (tag or "").strip().lower()
    if not lowered:
        return False
    for item in items:
        if lowered in {value.lower() for value in item.tags}:
            return True
    return False


def _target_for_context(ctx: PrerequisiteContext) -> EntityState | None:
    if ctx.target is not None:
        return ctx.target
    if ctx.nearby_entities:
        return ctx.nearby_entities[0]
    return None


def evaluate_prerequisite(prereq: Prerequisite, ctx: PrerequisiteContext) -> tuple[bool, str]:
    kind = prereq.kind.strip().lower()
    key = prereq.key.strip()
    reason = prereq.description or f"{prereq.kind}:{prereq.key}"

    if kind == "room_feature_is":
        return (ctx.room.feature == str(prereq.value), reason)
    if kind == "room_has_item_tag":
        return (ctx.room.has_item_tag(key), reason)
    if kind == "room_lacks_item_tag":
        return (not ctx.room.has_item_tag(key), reason)
    if kind == "exits_include":
        lowered = key.lower()
        return (lowered in {value.lower() for value in ctx.exits}, reason)
    if kind == "min_attribute":
        current = int(getattr(ctx.actor.attributes, key, 0))
        return (current >= int(prereq.value or 0), reason)
    if kind == "min_trait":
        current = float(ctx.actor.traits.values.get(key, 0.0))
        return (current >= float(prereq.value or 0.0), reason)
    if kind == "min_feature":
        current = float(ctx.actor.features.values.get(key, 0.0))
        return (current >= float(prereq.value or 0.0), reason)
    if kind == "has_inventory_tag":
        return (_inventory_has_tag(ctx.actor.inventory, key), reason)
    if kind == "target_exists":
        return (_target_for_context(ctx) is not None, reason)
    if kind == "target_has_item_tag":
        target = _target_for_context(ctx)
        if target is None:
            return (False, reason)
        return (_inventory_has_tag(target.inventory, key), reason)
    if kind == "skill_unlocked":
        skill_state = ctx.actor.skills.get(key)
        return (skill_state is not None and bool(skill_state.unlocked), reason)
    if kind == "min_level":
        return (ctx.actor.level >= int(prereq.value or 0), reason)
    if kind == "flag_true":
        return (bool(ctx.extra_flags.get(key, False)), reason)

    # Unknown prerequisite kinds are treated as blocked to avoid silent bypasses.
    return (False, f"unknown_prerequisite:{prereq.kind}")


def evaluate_prerequisites(
    prerequisites: Iterable[Prerequisite],
    ctx: PrerequisiteContext,
) -> AvailabilityResult:
    blocked: list[str] = []
    for prereq in prerequisites:
        passed, reason = evaluate_prerequisite(prereq, ctx)
        if not passed:
            blocked.append(reason)
    return AvailabilityResult(available=not blocked, blocked_reasons=tuple(blocked))

File: packages/engine/src/escape-the-dungeon/narrative/prerequisites.ts
import type { EntityState, RoomNode } from "../core/types";
import { hasRoomItemTag } from "../world/map";

export type PrerequisiteKind =
  | "room_feature_is"
  | "min_attribute"
  | "min_trait"
  | "min_feature"
  | "skill_unlocked"
  | "target_exists"
  | "target_has_item_tag"
  | "min_reputation"
  | "max_reputation"
  | "faction_is"
  | "target_faction_is"
  | "trait_below";

export interface Prerequisite {
  kind: PrerequisiteKind;
  key?: string;
  value?: number | string;
  description?: string;
}

export interface PrerequisiteContext {
  actor: EntityState;
  room: RoomNode;
  nearbyEntities: EntityState[];
  targetEntity?: EntityState | null;
}

export interface AvailabilityResult {
  available: boolean;
  blockedReasons: string[];
}

const selectTarget = (ctx: PrerequisiteContext): EntityState | null => {
  if (ctx.targetEntity) {
    return ctx.targetEntity;
  }
  return ctx.nearbyEntities[0] ?? null;
};

const attributeValue = (actor: EntityState, key: string): number => {
  return Number((actor.attributes as unknown as Record<string, number>)[key] ?? 0);
};

const traitValue = (actor: EntityState, key: string): number => {
  return Number((actor.traits as unknown as Record<string, number>)[key] ?? 0);
};

const featureValue = (actor: EntityState, key: string): number => {
  return Number((actor.features as unknown as Record<string, number>)[key] ?? 0);
};

const evaluateOne = (
  prereq: Prerequisite,
  ctx: PrerequisiteContext,
): { pass: boolean; reason: string } => {
  const reason = prereq.description ?? prereq.kind;

  if (prereq.kind === "room_feature_is") {
    return { pass: ctx.room.feature === String(prereq.value ?? ""), reason };
  }

  if (prereq.kind === "min_attribute") {
    const key = String(prereq.key ?? "");
    const minimum = Number(prereq.value ?? 0);
    const current = attributeValue(ctx.actor, key);
    return { pass: current >= minimum, reason };
  }

  if (prereq.kind === "min_trait") {
    const key = String(prereq.key ?? "");
    const minimum = Number(prereq.value ?? 0);
    return { pass: traitValue(ctx.actor, key) >= minimum, reason };
  }

  if (prereq.kind === "min_feature") {
    const key = String(prereq.key ?? "");
    const minimum = Number(prereq.value ?? 0);
    return { pass: featureValue(ctx.actor, key) >= minimum, reason };
  }

  if (prereq.kind === "skill_unlocked") {
    const key = String(prereq.key ?? "");
    return { pass: Boolean(ctx.actor.skills[key]?.unlocked), reason };
  }

  if (prereq.kind === "target_exists") {
    return { pass: selectTarget(ctx) !== null, reason };
  }

  if (prereq.kind === "target_has_item_tag") {
    const target = selectTarget(ctx);
    const tag = String(prereq.key ?? "");
    if (!target) {
      return { pass: false, reason };
    }
    return {
      pass: target.inventory.some((item) => item.tags.includes(tag)),
      reason,
    };
  }

  if (prereq.kind === "min_reputation") {
    return { pass: ctx.actor.reputation >= Number(prereq.value ?? 0), reason };
  }

  if (prereq.kind === "max_reputation") {
    return { pass: ctx.actor.reputation <= Number(prereq.value ?? 0), reason };
  }

  if (prereq.kind === "faction_is") {
    return {
      pass: ctx.actor.faction.toLowerCase() === String(prereq.value ?? "").toLowerCase(),
      reason,
    };
  }

  if (prereq.kind === "target_faction_is") {
    const target = selectTarget(ctx);
    if (!target) {
      return { pass: false, reason };
    }
    return {
      pass: target.faction.toLowerCase() === String(prereq.value ?? "").toLowerCase(),
      reason,
    };
  }

  if (prereq.kind === "trait_below") {
    const key = String(prereq.key ?? "");
    const threshold = Number(prereq.value ?? 0);
    return { pass: traitValue(ctx.actor, key) <= threshold, reason };
  }

  return { pass: false, reason: `unknown_prerequisite:${prereq.kind}` };
};

export const evaluatePrerequisites = (
  prerequisites: readonly Prerequisite[],
  ctx: PrerequisiteContext,
): AvailabilityResult => {
  const blockedReasons: string[] = [];
  for (const prereq of prerequisites) {
    const result = evaluateOne(prereq, ctx);
    if (!result.pass) {
      blockedReasons.push(result.reason);
    }
  }
  return {
    available: blockedReasons.length === 0,
    blockedReasons,
  };
};

export const roomHasItemTag = (room: RoomNode, tag: string): boolean => {
  return hasRoomItemTag(room, tag);
};
